File: 백준/Silver/17626. Four Squares/Four Squares.py
import sys

input= sys.stdin.readline

dp=[0,1]
n = int(input())
for i in range(2,n+1):
    min_value = 1e9
    j= 1

    while (j**2) <=i:
        min_value = min(min_value,dp[i-(j**2)])
        j+=1

    dp.append(min_value+1)
print(dp[n])

File: 백준/Silver/17626. Four Squares/Four Squares.rs
use std::io::{BufReader,BufRead,BufWriter,Write,stdin,stdout};
fn main(){
    let mut reader= BufReader::new(stdin().lock());
    let mut writer=BufWriter::new(stdout().lock());
    let mut input = String::new();
    reader.read_line(&mut input).unwrap();
    let n= input.trim().parse().unwrap();

 
    let mut dp: Vec<usize> = vec![0, 1];

    for i in 2..=n {
        /*10의 9승 */
        let mut min_value = 1e9 as usize;
        let mut j = 1;

        while j * j <= i {
            /*비교 */
            min_value = min_value.min(dp[i - j * j]);
            j += 1;
        }

        dp.push(min_value + 1);
    }
    writeln!(writer,"{}",dp[n]).unwrap();
    writer.flush().unwrap();

}